### ISSUE TITLE
Serverside rendering 

### ISSUE THREAD

**Issue by ennair** [bloomreach/spa-sdk#38]
Hi, 
In the current example of Next, the highest component BrxApp (https://github.com/bloomreach/spa-sdk/blob/main/examples/next/components/BrxApp.tsx#L1) is a clientside component, because of the use of BrPage which is a client component. All components underneath therefore will also be clientside components, also because they need to get the page and component contexts. Is it possible to set it up serverside? So we can render our components more serverside? There are a couple of benefits of rendering it serverside: https://nextjs.org/docs/app/building-your-application/rendering/server-components#benefits-of-server-rendering.

**Comment by ksalic** [comment]
@hachok would you be able to check this?

**Comment by ennair** [comment]
I get what you mean, only this is not working for the mapping components which you provide to BrPage. If you add components via children of BrPage you will be able to run them serverside, but then the question remains how to get the page data without context. Because when using context it will have to be a client component again. I have made a small example. 

app/[[...route]]/page.tsx
```
import BrAllPages from '../../components/BrAllPages';
import { getBrPageModel } from '../../utils/https-utils';
import { headers } from 'next/headers';

export default async function Page({ searchParams }) {
  const { ...query } = searchParams;

  const headersList = headers();
  const pathname = headersList.get('x-pathname');

  const { page, ...props } = await getBrPageModel(
    pathname,
    'nl',
    query,
    headersList,
    {},
  );

  return <BrAllPages configuration={props} page={page} />;
}
```

components/BrAllPages/BrAllPages.tsx (serverside component)
```
import BrPageWrapper from '../../components/layout/BrPageWrapper';
import BrTextViaChildren from '../BrTextViaChildren';

export default function AllPages({ configuration, page }) {
  console.log('BrAllPages');

  return (
    <BrPageWrapper page={page} configuration={configuration}>
      <BrTextViaChildren />
    </BrPageWrapper>
  );
}
```

As you can see I gave the BrPageWrapper component (clientside component) a child BrTextViaChildren component (serverside component). This works, but how do I get the page data from BrPageWrapper without using context? Because when I start using context in BrTextViaChildren it has to be a clientside component.

components/BrTextViaChildren/BrTextViaChildren (serverside component)
```
export function BrTextViaChildren() {
  console.log('BrTextChildren');

  return (
    <div className='component-wrapper content-small-wrapper'>
      <p>Text via children, but how?</p>
    </div>
  );
}
```

components/BrPageWrapper/BrPageWrapper (clientside component)
```
'use client';

import { BrComponent, BrPage } from '@bloomreach/react-sdk';
import axios from 'axios';
import BrTextViaMapping from '../BrTextViaMapping';

type BrPageWrapperProps = {
  page: any;
  configuration: any;
  children: any;
  metaData?: any;
};

export function BrPageWrapper({
  page,
  configuration,
  children,
}: BrPageWrapperProps) {
  console.log('BrPageWrapper');

  const mapping = {
    'Inhoud': BrTextViaMapping,
  };

  return (
    <div id='p1o-app' className='page-wrapper'>
      <BrPage
      configuration={{ ...configuration, httpClient: axios }}
      mapping={mapping}
      page={page}
    >
      <BrComponent path='main-menu' />

      <BrComponent path='information-area' />
      {children}

    </BrPage>
    </div>
  );
}
```

components/BrTextViaMapping/BrTextViaMapping (because it is used in via de mapping of BrPage it automatically becomes a clientside component)
```
import { BrProps } from '@bloomreach/react-sdk';
import { Content, Page } from '@bloomreach/spa-sdk';

export function BrTextViaMapping(props: BrProps) {
  const page: Page = props.page;
  const document: Content = page?.getDocument();

  console.log('BrTextViaMapping');

  const content = document?.getData()?.introduction?.value;

  return (
    <div className='component-wrapper content-small-wrapper'>
      {content && <p>{content}</p>}
    </div>
  );
}
```

I tried it via the children, but then I do not know how to get the page data from the BrPage. I also have tried it via the mapping, then I am able to get the page data, but the component automatically becomes a clientside component and not serverside component. 

So how am I able to make a component, which can access the Bloomreach page data and is serverside rendered?

**Comment by ennair** [comment]
Yes, but still a part of it is still done on the client as you can also see in the links you send. I mean for components without any interactivity needed, like RichText components, of which a page often has the most of, wouldn't it be nicer if we could have them as server components instead of client components?

**Comment by cwmrowe** [comment]
Hi @hachok while we wait for React Server Components support, is there documentation of how to use Bloomreach CMS in a pure NodeJS environment? i.e  one that does not require JS in the browser?

**Comment by cwmrowe** [comment]
@hachok any update on the support for React Server Components, or details on how we can use Bloomreach with server side rendered pages in NodeJS. This is preventing our upgrade from NextJS pages to NextJS App Router. https://nextjs.org/docs/app

**Comment by joerideg** [comment]
Hi @cwmrowe ,
We have raised this internally and we are planning to rewrite the SDK into functional components as class components are not supported in RSC. Then we will also take a look at changing the way the context is used, such that you can use the context separately from BrPage and thus align more with RSC usage. We will post updates here as we make progress.

**Comment by cwmrowe** [comment]
Hi @joerideg. One of the benefits of RSCs, or indeed server rendering in general, is that we need fewer JavaScript dependencies in the browser. 

Will you consider the option of not requiring Bloomreach JS in the browser for RSCs? I appreciate some JS is needed for Experience Manager when in preview mode, but it would be better for performance if we didn't need to include Bloomreach JS for our end users. 

I am happy to help with the design of this.

**Comment by joerideg** [comment]
Hi @cwmrowe ,
I think we are on the same page. Its what I intended to convey: We are looking to support RSC with the React SDK, which means we will write them as such as much as possible. For that we also need to convert our current component syntax, class components, to functional components. I dont see any reason why supporting RSC would not be possible considering we are really just rendering out a component hierarchy of static templates after a single fetch of the model from the backend. Your mapped components might or might not be a RSC, but React should probably pick that up automatically. 

What we are still wondering about is how we would be able to expose the Page object to be used somewhere in the hierarchy, for example to render out the page title or something small that would not require a component. I'm sure we will find a way to do so with context without forcing the entire component tree to be client components.

**Comment by ulrichgreen** [comment]
I dont have anything to add, other than our excitement as another SDK user to get RSC support soon! :)

**Comment by lachire** [comment]
Hi @cwmrowe ,

I am happy to announce that a new version [26.0.0](https://github.com/bloomreach/spa-sdk/releases/tag/spa-sdk-26.0.0) has been released with the support for React Server Components. We removed all React Context APIs, so this is a breaking change. Please follow the release notes on how to migrate to the new version. Thanks!

**Comment by ennair** [comment]
Hi @lachire , thanks for the latest update!

I’ve been testing version 26.0.0 and noticed that [BrPage](https://github.com/bloomreach/spa-sdk/blob/spa-sdk-26.0.0/packages/react-sdk/src/page/BrPage.tsx) remains a client-side component due to its use of useState and useEffect. As a result, in your example [BrxApp.tsx](https://github.com/bloomreach/spa-sdk/blob/spa-sdk-26.0.0/examples/next/components/BrxApp.tsx), BrxApp is also rendered as a client component.

Would it be possible to refactor BrPage into a server-side component? That would allow us to structure the rest of the application more server-side, aligning better with Next.js's app directory and server component architecture.

Thanks again for your work on this!

**Comment by joerideg** [comment]
@ennair thanks for trying it out and giving feedback! I'm taking another look at this. Will get back to you soon.

**Comment by lachire** [comment]
Hi @ennair , I am happy to announce that a new version [27.0.0](https://github.com/bloomreach/spa-sdk/releases/tag/spa-sdk-27.0.0) has been released with full support for React Server Components. 

For backwards compatibility, the default rendering mode will still be using client components. Significant changes might be needed if you want to adopt server components. Please check the `README` file as well as the `example/next` app for implementation details/examples.

**Comment by beetlerom** [comment]
Since no feedback was shared for a while, I am closing as this issue. Please open a new ticket should there be any further feedback.

**Comment by ennair** [comment]
Thank you for the change! I was busy trying it out. Looks good! I will let you know if I have any feedback.
